Add WebSession auth option and support for background scripts, GlideAjax, sn_cicd and sn_devstudio APIs.

dimitertodorov opened the pull request:
This pull request covers off a number of features I have been using for a while now. I refactored them to fit the into the PSSnow module. I have included UNIT tests, as well as Integration tests.


### Features

- Authentication using WebSessions and the ServiceNow `X-UserToken` header.
- Support for running background scripts using `Invoke-SNOWBackgroundScript`
- Support for running GlideAjax requests using `Invoke-SNOWGlideAjax` and `Wait-SNOWGlideAjaxProgress`.
- Support for `sn_cicd` and `sn_devstudio` API endpoints for managing update sets and VCS applications.

### Changed

- [`Set-SNOWAuth`](src/Public/Set-SNOWAuth.ps1)
  - Added `-UseWebSession` parameter to start a new WebSession and use it for authentication. This works with all authentication methods (Basic, OAuth, OAuthToken).
  - Added `OAuthToken` ParameterSet with new parameters `-AccessToken` and `-RefreshToken`  for OAuth authentication using an existing access token and refresh token.
  
- [`Assert-SNOWAuth`](src/Private/Assert-SNOWAuth.ps1)
  - Update OAuth refresh flow. ClientSecret is now optional and will be used if provided. If not, a refresh will be attempted using the existing `$script:SNOWAuth.token.refresh_token` and `$script:SNOWAuth.ClientId` only.
  - When a WebSession is used, the function will now check if the cookies are still valid and refreshes it if necessary. This sets the `$Script:SNOWAuth.SessionState` variable using the new `Get-SNOWWebSessionState` function.

  [`Invoke-SNOWWebRequest`](src/Public/Invoke-SNOWWebRequest.ps1)
  - Update to use the `$Script:SNOWAuth.session.WebSession` and `$script:SNOWAuth.SessionState.SecurityToken` variables for authentication and session management when available and valid.

### Added

#### [web](src/Public/web/)

- [`Get-SNOWWebConcourseState`](src/Public/web/Get-SNOWWebConcourseState.ps1) - Retrieves the current state of the active ServiceNow session.
- [`Get-SNOWWebSessionState`](src/Public/web/Get-SNOWWebSessionState.ps1) - Retrieves the current state of a ServiceNow web session.
- [`Invoke-SNOWBackgroundScript`](src/Public/web/Invoke-SNOWBackgroundScript.ps1) - Executes a background script in the ServiceNow instance.
- [`Invoke-SNOWGlideAjax`](src/Public/web/Invoke-SNOWGlideAjax.ps1) - Sends a GlideAjax request to a ServiceNow instance and returns the response synchronously.
- [`New-SNOWAuthWebSession`](src/Public/web/New-SNOWAuthWebSession.ps1) - Creates an authenticated web session to ServiceNow using credentials from SNOWAuth.
- [`Set-SNOWWebConcourseState`](src/Public/web/Set-SNOWWebConcourseState.ps1) - Sets the current application or update set in a ServiceNow web session.
- [`Wait-SNOWGlideAjaxProgress`](src/Public/web/Wait-SNOWGlideAjaxProgress.ps1) - Waits for a ServiceNow background process to complete.

#### [sn_cicd](src/Public/sn_cicd/)

- [`Get-SNOWUpdateSet`](src/Public/sn_cicd/Get-SNOWUpdateSet.ps1) - Get update set from ServiceNow by sys_id. Searches sys_update_set and sys_remote_update_set tables in order.
- [`Import-SNOWUpdateSet`](src/Public/sn_cicd/Import-SNOWUpdateSet.ps1) - Imports an update set XML file into ServiceNow using sys_upload.do and a valid WebSession.
- [`New-SNOWUpdateSet`](src/Public/sn_cicd/New-SNOWUpdateSet.ps1) - Creates a new update set in ServiceNow using the CICD API.
- [`Remove-SNOWUpdateSet`](src/Public/sn_cicd/Remove-SNOWUpdateSet.ps1) - Removes a ServiceNow update set by directly deleting it.
- [`Search-SNOWUpdateSet`](src/Public/sn_cicd/Search-SNOWUpdateSet.ps1) - Searches for update sets in ServiceNow.
- [`Start-SNOWUpdateSetBackOut`](src/Public/sn_cicd/Start-SNOWUpdateSetBackOut.ps1) - Removes (backs out) an update set from ServiceNow.
- [`Start-SNOWUpdateSetCommit`](src/Public/sn_cicd/Start-SNOWUpdateSetCommit.ps1) - Starts the commit process for a ServiceNow update set.
- [`Start-SNOWUpdateSetPreview`](src/Public/sn_cicd/Start-SNOWUpdateSetPreview.ps1) - Starts the preview process for a ServiceNow update set.
- [`Start-SNOWVCSApplyChanges`](src/Public/sn_cicd/Start-SNOWVCSApplyChanges.ps1) - Starts applying changes from a remote source control to a specified local application or application-customization.
- [`Start-SNOWVCSApplyStash`](src/Public/sn_cicd/Start-SNOWVCSApplyStash.ps1) - Starts applying a previously generated "stash" of changes from a remote source control.
- [`Start-SNOWVCSImport`](src/Public/sn_cicd/Start-SNOWVCSImport.ps1) - Imports an application using the specified repository URL and branch name.
- [`Sync-SNOWVCSApplication`](src/Public/sn_cicd/Sync-SNOWVCSApplication.ps1) - Synchronizes a ServiceNow application with a Git repository using the CICD Source Control API.
- [`Test-SNOWUpdateSet`](src/Public/sn_cicd/Test-SNOWUpdateSet.ps1) - Tests an update set XML file to determine if it contains an update set.
- [`Wait-SNOWCICDProgress`](src/Public/sn_cicd/Wait-SNOWCICDProgress.ps1) - Waits for a ServiceNow CICD operation to complete.

#### [sn_devstudio](src/Public/sn_devstudio/)

- [`Get-SNOWDevStudioApp`](src/Public/sn_devstudio/Get-SNOWDevStudioApp.ps1) - Get all VCS apps from api/sn_devstudio/v1/vcs/apps
- [`Sync-SNOWDevStudioApp`](src/Public/sn_devstudio/Sync-SNOWDevStudioApp.ps1) - Imports, syncs and updates a ServiceNow application with a Git repository. Use sn_devstudio API endpoints to manage the repository.
- [`Wait-SNOWDevStudioTransaction`](src/Public/sn_devstudio/Wait-SNOWDevStudioTransaction.ps1) - Wait for a sn_devstudio transaction to complete - api/sn_devstudio/v1/vcs/transactions/{ProgressId}

dimitertodorov committed:
Make ClientSecret optional for refresh token flow to support public OAuth Client token.

dimitertodorov committed:
Add WebSession authentication option, background script execution, GlideAjax execution, and Public OAuth Client token support.

### Features

- Authentication using WebSessions and the ServiceNow `X-UserToken` header.
- Support for running background scripts using `Invoke-SNOWBackgroundScript`
- Support for running GlideAjax requests using `Invoke-SNOWGlideAjax` and `Wait-SNOWGlideAjaxProgress`.

### Changed

- [`Set-SNOWAuth`](src/Public/Set-SNOWAuth.ps1)
  - Added `-UseWebSession` parameter to start a new WebSession and use it for authentication. This works with all authentication methods (Basic, OAuth, OAuthToken).
  - Added `OAuthToken` ParameterSet with new parameters `-AccessToken` and `-RefreshToken`  for OAuth authentication using an existing access token and refresh token.

- [`Assert-SNOWAuth`](src/Private/Assert-SNOWAuth.ps1)
  - Update OAuth refresh flow. ClientSecret is now optional and will be used if provided. If not, a refresh will be attempted using the existing `$script:SNOWAuth.token.refresh_token` and `$script:SNOWAuth.ClientId` only.
  - When a WebSession is used, the function will now check if the cookies are still valid and refreshes it if necessary. This sets the `$Script:SNOWAuth.SessionState` variable using the new `Get-SNOWWebSessionState` function.

  [`Invoke-SNOWWebRequest`](src/Public/Invoke-SNOWWebRequest.ps1)
  - Update to use the `$Script:SNOWAuth.session.WebSession` and `$script:SNOWAuth.SessionState.SecurityToken` variables for authentication and session management when available and valid.

### Added

#### [web](src/Public/web/)

- [`Get-SNOWWebConcourseState`](src/Public/web/Get-SNOWWebConcourseState.ps1) - Retrieves the current Application and Update Set of the active ServiceNow session.
- [`Get-SNOWWebSessionState`](src/Public/web/Get-SNOWWebSessionState.ps1) - Retrieves the current state of a ServiceNow web session.
- [`Invoke-SNOWBackgroundScript`](src/Public/web/Invoke-SNOWBackgroundScript.ps1) - Executes a background script in the ServiceNow instance.
- [`Invoke-SNOWGlideAjax`](src/Public/web/Invoke-SNOWGlideAjax.ps1) - Sends a GlideAjax request to a ServiceNow instance and returns the response synchronously.
- [`New-SNOWAuthWebSession`](src/Public/web/New-SNOWAuthWebSession.ps1) - Creates an authenticated web session to ServiceNow using credentials from SNOWAuth.
- [`Set-SNOWWebConcourseState`](src/Public/web/Set-SNOWWebConcourseState.ps1) - Sets the current application or update set in a ServiceNow web session.
- [`Wait-SNOWGlideAjaxProgress`](src/Public/web/Wait-SNOWGlideAjaxProgress.ps1) - Waits for a ServiceNow background process to complete.

dimitertodorov committed:
Update generated docs

dimitertodorov committed:
Support for `sn_cicd` and `sn_devstudio` API endpoints for managing update sets and VCS applications.

#### [sn_cicd](src/Public/sn_cicd/)

- [`Get-SNOWUpdateSet`](src/Public/sn_cicd/Get-SNOWUpdateSet.ps1) - Get update set from ServiceNow by sys_id. Searches sys_update_set and sys_remote_update_set tables in order.
- [`Import-SNOWUpdateSet`](src/Public/sn_cicd/Import-SNOWUpdateSet.ps1) - Imports an update set XML file into ServiceNow using sys_upload.do and a valid WebSession.
- [`New-SNOWUpdateSet`](src/Public/sn_cicd/New-SNOWUpdateSet.ps1) - Creates a new update set in ServiceNow using the CICD API.
- [`Remove-SNOWUpdateSet`](src/Public/sn_cicd/Remove-SNOWUpdateSet.ps1) - Removes a ServiceNow update set by directly deleting it.
- [`Search-SNOWUpdateSet`](src/Public/sn_cicd/Search-SNOWUpdateSet.ps1) - Searches for update sets in ServiceNow.
- [`Start-SNOWUpdateSetBackOut`](src/Public/sn_cicd/Start-SNOWUpdateSetBackOut.ps1) - Removes (backs out) an update set from ServiceNow.
- [`Start-SNOWUpdateSetCommit`](src/Public/sn_cicd/Start-SNOWUpdateSetCommit.ps1) - Starts the commit process for a ServiceNow update set.
- [`Start-SNOWUpdateSetPreview`](src/Public/sn_cicd/Start-SNOWUpdateSetPreview.ps1) - Starts the preview process for a ServiceNow update set.
- [`Start-SNOWVCSApplyChanges`](src/Public/sn_cicd/Start-SNOWVCSApplyChanges.ps1) - Starts applying changes from a remote source control to a specified local application or application-customization.
- [`Start-SNOWVCSApplyStash`](src/Public/sn_cicd/Start-SNOWVCSApplyStash.ps1) - Starts applying a previously generated "stash" of changes from a remote source control.
- [`Start-SNOWVCSImport`](src/Public/sn_cicd/Start-SNOWVCSImport.ps1) - Imports an application using the specified repository URL and branch name.
- [`Sync-SNOWVCSApplication`](src/Public/sn_cicd/Sync-SNOWVCSApplication.ps1) - Synchronizes a ServiceNow application with a Git repository using the CICD Source Control API.
- [`Test-SNOWUpdateSet`](src/Public/sn_cicd/Test-SNOWUpdateSet.ps1) - Tests an update set XML file to determine if it contains an update set.
- [`Wait-SNOWCICDProgress`](src/Public/sn_cicd/Wait-SNOWCICDProgress.ps1) - Waits for a ServiceNow CICD operation to complete.

#### [sn_devstudio](src/Public/sn_devstudio/)

- [`Get-SNOWDevStudioApp`](src/Public/sn_devstudio/Get-SNOWDevStudioApp.ps1) - Get all VCS apps from api/sn_devstudio/v1/vcs/apps
- [`Sync-SNOWDevStudioApp`](src/Public/sn_devstudio/Sync-SNOWDevStudioApp.ps1) - Imports, syncs and updates a ServiceNow application with a Git repository. Use sn_devstudio API endpoints to manage the repository.
- [`Wait-SNOWDevStudioTransaction`](src/Public/sn_devstudio/Wait-SNOWDevStudioTransaction.ps1) - Wait for a sn_devstudio transaction to complete - api/sn_devstudio/v1/vcs/transactions/{ProgressId}

dimitertodorov committed:
Update CHANGELOG.md with new unreleased functionality.

dimitertodorov committed:
Add missing functions to PSSnow.psd1

dimitertodorov committed:
Remove development proxy reference in test helper.

dimitertodorov commented:
@insomniacc - Could you take a look and see if the changes fit with your vision for this Module.

dimitertodorov committed:
Fix up tests to ignore certain PSScriptAnalyzer rules.

insomniacc commented:
@dimitertodorov  Hey man! Awesome work, I really appreciate you putting time and effort into improving the module, I've not had a great deal of time myself this past year to add to it. I'll take a look properly at some point this week and go through it all. Will let you know what I think!

insomniacc commented on src/Public/web/New-SNOWAuthWebSession.ps1:
```diff
@@ -0,0 +1,167 @@
+        # Alternative Login Method
+        # Form-Based login is left here for backward compatibility in case the first method fails
+        $LoginParams = @{
+            Uri        = "https://$Instance.service-now.com/login.do"
```
Where is $instance defined? I'm not seeing it as a parameter and the variable isnt being scoped to $Script either?

insomniacc commented on src/Public/web/Invoke-SNOWBackgroundScript.ps1:
```diff
@@ -0,0 +1,64 @@
+    
+    $Session = Get-SNOWWebSessionState -ValidateSession
+    if (-not $Session.Valid) {
+        Write-Error "No active ServiceNow session found. Please connect using Connect-SNOW first."
```
Connect-SNOW doesnt exist was this a typo/mistake? Should it have been Set-SnowAuth?
Also, is $Session required here for this function? could I not trigger a background script for example, if just using OAuth authentication without the -UseWebSession switch? I havent looked into the endpoint much but I'm assuming this is because it's not a typical API endpoint as such and the $session is needed for that reason?

I'd suggest handling it in the way it's done in Invoke-SNOWGlideAjax lines 37 - 39.

insomniacc commented on src/Public/Set-SNOWAuth.ps1:
```diff
@@ -29,40 +40,66 @@ function Set-SNOWAuth {
+        $UseWebSession,
+        [Parameter(Mandatory, ParameterSetName = 'OAuthToken')]
+        [string]
+        # An e
```
Missing the rest of the sentence

insomniacc commented on src/Public/web/New-SNOWAuthWebSession.ps1:
```diff
@@ -0,0 +1,167 @@
+        $EncodedPassword = [uri]::EscapeDataString($Credential.GetNetworkCredential().Password)
+        # Perform login request
+        $AuthParams = @{
+            Uri        = "https://$Instance.service-now.com/login.do?user_name=$($Credential.UserName)&user_password=${EncodedPassword}&sys_action=sysverb_login&sysparm_ck=$SecurityToken&sysparm_goto_url=/sys_user_list.do?sysparm_limit=1&sysparm_fields=name,sys_id,user_name"
```
$Instance & Credential are used again here, should it have been scoped to $script:SNOWAuth.Instance ?

insomniacc reviewed (changes requested):
Just a couple of comments/minor details, other than that I'm happy - this all looks like really good additions, so thanks again for the effort. Let me know what you think on what I've pointed out / correct me if I've got anything wrong and I'll merge the changes and go from there!

insomniacc commented on src/Public/web/New-SNOWAuthWebSession.ps1:
```diff
@@ -0,0 +1,167 @@
+        }
+            
+        # URL encode the password for the login request
+        $EncodedPassword = [uri]::EscapeDataString($Credential.GetNetworkCredential().Password)
```
I think credential is also not defined/scoped to the $Script variable?

dimitertodorov commented:
Hey - Thanks you for all the feedback. This was exactly what I was looking for someone to do since I am mostly working solo on this project.

Will fix these up along with some other errors I encountered.

During recent testing I also found that the sn_devstudio endpoint returns 403 for non-admin users.
When using Basic auth this is fine since we fallback to traditional login.do session login.
This fallback method does not work in an OAuth context - so I will raise a clearer error, along with the required role needed to call that endpoint.

dimitertodorov committed:
Mix missing parameter documentation

dimitertodorov committed:
Fix missing references for instance and credential

dimitertodorov committed:
Add Assert-SNOWAuthWebSession

dimitertodorov committed:
Catch 401 errors when using a WebSession by re-authenticating and retrying the request.

dimitertodorov committed:
Add -ForceCommit to Start-SNOWUpdateSetCommit.ps1

dimitertodorov committed:
doc fixes

dimitertodorov committed:
Invoke-SNOWBackgroundScript will now try to create a new session if one does not exist in the current context.

dimitertodorov committed:
Fix websession integration test.

dimitertodorov committed:
Handle 'global' and globally scoped applications separately.

dimitertodorov committed:
Use Assert-SNOWAuthWebSession when importing update set from XML.

dimitertodorov committed:
Various fixes to VCS Application sync. Handle globally scoped applications properly now.

dimitertodorov committed:
Use Assert-SNOWAuthWebSession when running background scripts
When specifying a scope, we now check the current application set for a given scope to make sure it exists.

dimitertodorov committed:
Fix SourceControl integration tests

dimitertodorov committed:
Merge pull request #1 from cherichita/1.4.0

Various fixes before merge into PSSnow Upstream

dimitertodorov commented:
Need to rethink the web-session handling. Will close this for now while I do a bit of digging.
Latest Yokohama release no longer accepts g_ck tokens from the sn_devstudio_/v1/get_publish_info endpoint when running background scripts.
